### DISCUSSION TITLE
How can I parse tokens but ignore some occurrences?

### DISCUSSION THREAD

**Discussion by felixjb** [babel/babel#11961] [Q&A] [answered]
I am using the @babel/parser to find the testing framework keywords and add VSCode codelens on them. However, it gives a false positive when the code contains a regex test method.

I would like to find the word "test" for the Jest and Mocha frameworks, but not for a regex test method.

Here's the code where do the parsing: [Testify: code parser](https://github.com/felixjb/testify/blob/main/src/parser/codeParser.ts)

**Comment by kaicataldo** [comment, marked as answer]
In the context of tokens, you could potentially check the previous token to see if it's a `.` or not. Otherwise, I think this would be easier in the context of the AST, as you can specifically check if the call is a `MemberExpression` or not.

**Reply by felixjb** [reply]
How would I do that? Can I do that in the `.map()` function in my [code](https://github.com/felixjb/testify/blob/f924b3185c3a558d61b4f66baa16ec0e8680b4c7/src/parser/codeParser.ts#L27)?

**Reply by kaicataldo** [reply]
Can you clarify which option you're asking about? I would recommend traversing the AST rather than iterating over the tokens, since you already have it there.

**Reply by felixjb** [reply]
Sorry, I'll clarify.

Today I am iterating over the tokens to find occurrences of test keywords from unit testing frameworks. When I find them, I return their position.

How could I replace this implementation with traversing the AST?

**Reply by kaicataldo** [reply]
Here's generally how I would approach this (I haven't done this specifically, so you might have to do some tweaking):

1. Traverse the AST, visiting each node starting from the root
2. When you encounter a node with type `CallExpression`, check the `callee` property, verify that it's `type: Identifier`, and check the `name` value.
3. Use the location of the `Identifier` node.

**Reply by felixjb** [reply]
Nice! I'm gonna try this then.

**Reply by kaicataldo** [reply]
Forgot to mention, https://astexplorer.net/ is really useful for visualizing the AST!
